
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="../../../_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="../../../_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="../../../_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="../../../_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>tridentplus.organisms.genome &#8212; TridentPlus 0.0.1 documentation</title>
    <link rel="stylesheet" href="../../../_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/custom.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  
   

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">TridentPlus 0.0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar"><a href="" class="github-corner" aria-label="View source on GitHub"><svg width="80" height="80" viewBox="0 0 250 250" style="fill:#fd8ea4; color:#fff; position: absolute; top: 0; border: 0; left: 0; transform: scale(-1, 1);" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>

<div>
    <div>
        <h1 align="center"><a href="../../../index.html">Trident<font color="black">+</font></a>
        </h1>
    </div>
</div>

<p align="center" class="sidebar-subtitle">Trident+ version: <a href="">0.0.1</a></p>
<div>

</div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Table Of Contents</h2>
  </div>
  <div class="sidebar-toc">
    
    
      <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Trident+ Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../developer/api_reference.html">API Reference</a></li>
</ul>

    
  </div>
</div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="../../../search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
      
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="../../../index.html">Docs</a></li>
              
                <li><a href="../../index.html">Module code</a></li>
              
              <li>tridentplus.organisms.genome</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <h1>Source code for tridentplus.organisms.genome</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Genome services</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">tridentplus.primers</span> <span class="k">import</span> <span class="n">pick_pcr_primers</span>
<span class="kn">from</span> <span class="nn">pydent</span> <span class="k">import</span> <span class="n">AqSession</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">product</span>

<span class="kn">import</span> <span class="nn">jdna</span>
<span class="kn">import</span> <span class="nn">pyblast</span>


<div class="viewcode-block" id="BlastUtils"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.BlastUtils">[docs]</a><span class="k">class</span> <span class="nc">BlastUtils</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Blast utilities&quot;&quot;&quot;</span>

<div class="viewcode-block" id="BlastUtils.align"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.BlastUtils.align">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">align</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">queries</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">subjects</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Performs an alignment between queries and subjects. Subject sequences will be collated into</span>
<span class="sd">        a single FASTA file and local blast installation will build a blastdb in a temporary</span>
<span class="sd">        directory and perform a query.</span>

<span class="sd">        :param queries: list of query sequences (as strings)</span>
<span class="sd">        :type queries: list</span>
<span class="sd">        :param subjects: list of subject sequences (as strings) to build a blastdb out of</span>
<span class="sd">        :type subjects:</span>
<span class="sd">        :return: aligner instance, with results available via `aligner.results` or `aligner.results.alignments`. Sequence</span>
<span class="sd">                    dictionary is available via `aligner.seq_dict`. Raw results available via `aligner.raw_results`.</span>
<span class="sd">        :rtype: pyblast.JSONBlast</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">query_json</span> <span class="o">=</span> <span class="p">[{</span>
            <span class="s1">&#39;bases&#39;</span><span class="p">:</span> <span class="n">q</span><span class="p">,</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;query_</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
            <span class="s1">&#39;circular&#39;</span><span class="p">:</span> <span class="kc">False</span>
        <span class="p">}</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">q</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">queries</span><span class="p">)]</span>

        <span class="n">subject_json</span> <span class="o">=</span> <span class="p">[{</span>
            <span class="s1">&#39;bases&#39;</span><span class="p">:</span> <span class="n">q</span><span class="p">,</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;subject_</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">),</span>
            <span class="s1">&#39;circular&#39;</span><span class="p">:</span> <span class="kc">False</span>
        <span class="p">}</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">q</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">subjects</span><span class="p">)]</span>

        <span class="n">aligner</span> <span class="o">=</span> <span class="n">pyblast</span><span class="o">.</span><span class="n">JSONBlast</span><span class="p">(</span><span class="n">query_json</span><span class="o">=</span><span class="n">query_json</span><span class="p">,</span> <span class="n">subject_json</span><span class="o">=</span><span class="n">subject_json</span><span class="p">,</span> <span class="n">span_origin</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">aligner</span><span class="o">.</span><span class="n">quick_blastn</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">aligner</span></div>

<div class="viewcode-block" id="BlastUtils.alignments_to_integration_sites"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.BlastUtils.alignments_to_integration_sites">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">alignments_to_integration_sites</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">alignments</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parse alignments (from :class:`pyblast.JSONBlast` instance) into integration sites.</span>

<span class="sd">        :param alignments: alignment list</span>
<span class="sd">        :type alignments: list</span>
<span class="sd">        :return: list of sites compose of tuple of (integration_direction, alignment1, alignment2)</span>
<span class="sd">        :rtype: list</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># group alignments by chromosome</span>
        <span class="n">group_by_subject_name</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
            <span class="n">group_by_subject_name</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>

        <span class="n">sites</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">subject_name</span><span class="p">,</span> <span class="n">aligns</span> <span class="ow">in</span> <span class="n">group_by_subject_name</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

            <span class="c1"># pair up alignments</span>
            <span class="n">pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">product</span><span class="p">(</span><span class="n">aligns</span><span class="p">,</span> <span class="n">aligns</span><span class="p">))</span>

            <span class="c1"># filter out pairs that use different queries</span>
            <span class="n">pairs</span> <span class="o">=</span> <span class="p">[(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">)</span> <span class="k">for</span> <span class="n">p1</span><span class="p">,</span> <span class="n">p2</span> <span class="ow">in</span> <span class="n">pairs</span> <span class="k">if</span> <span class="n">p1</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">p2</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]]</span>

            <span class="c1"># filter out pairs that have same start and end</span>
            <span class="n">pairs</span> <span class="o">=</span> <span class="p">[(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">)</span> <span class="k">for</span> <span class="n">p1</span><span class="p">,</span> <span class="n">p2</span> <span class="ow">in</span> <span class="n">pairs</span> <span class="k">if</span> <span class="n">p1</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">p2</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="ow">and</span>
                     <span class="n">p1</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">p2</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">]]</span>

            <span class="k">for</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">:</span>
                <span class="n">s1</span> <span class="o">=</span> <span class="n">a1</span><span class="p">[</span><span class="s1">&#39;subject&#39;</span><span class="p">]</span>
                <span class="n">s2</span> <span class="o">=</span> <span class="n">a2</span><span class="p">[</span><span class="s1">&#39;subject&#39;</span><span class="p">]</span>
                <span class="n">q1</span> <span class="o">=</span> <span class="n">a1</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">]</span>
                <span class="n">q2</span> <span class="o">=</span> <span class="n">a2</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">]</span>

                <span class="k">if</span> <span class="n">q1</span><span class="p">[</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">q2</span><span class="p">[</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]:</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">q1</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">q2</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]:</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">s1</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">s2</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]:</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">q1</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">q2</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]:</span>
                    <span class="k">continue</span>

                <span class="n">direction</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">s1_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">s1</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">s1</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]])</span>
                <span class="n">s1_min</span> <span class="o">=</span> <span class="nb">min</span><span class="p">([</span><span class="n">s1</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">s1</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]])</span>
                <span class="n">s2_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">s2</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">s2</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]])</span>
                <span class="n">s2_min</span> <span class="o">=</span> <span class="nb">min</span><span class="p">([</span><span class="n">s2</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">s2</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]])</span>

                <span class="k">if</span> <span class="n">s1</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;plus&#39;</span> <span class="ow">and</span> <span class="n">s1_max</span> <span class="o">&lt;=</span> <span class="n">s2_min</span><span class="p">:</span>
                    <span class="n">direction</span> <span class="o">=</span> <span class="n">s2_min</span> <span class="o">-</span> <span class="n">s1_max</span>
                <span class="k">elif</span> <span class="n">s1</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;minus&#39;</span> <span class="ow">and</span> <span class="n">s2_max</span> <span class="o">&lt;=</span> <span class="n">s1_min</span><span class="p">:</span>
                    <span class="n">direction</span> <span class="o">=</span> <span class="n">s2_max</span> <span class="o">-</span> <span class="n">s1_min</span>
                <span class="k">if</span> <span class="n">direction</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">sites</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">direction</span><span class="p">,</span> <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">sites</span></div>

<div class="viewcode-block" id="BlastUtils.aligner_to_integrations"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.BlastUtils.aligner_to_integrations">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">aligner_to_integrations</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">aligner</span><span class="p">,</span> <span class="n">integrant</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">template</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_homology</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flanking_bps</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
                                 <span class="n">max_distance</span><span class="o">=</span><span class="mi">5000</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parse a **run** :class:`pyblast.JSONBlast` instance into valid integration sites. Returns the following info</span>
<span class="sd">        for each integration:</span>

<span class="sd">        .. code-block::</span>

<span class="sd">            integration_info = {</span>
<span class="sd">                &#39;subject_name&#39;: site[1][&#39;subject&#39;][&#39;name&#39;],</span>
<span class="sd">                &#39;query_name&#39;: site[1][&#39;query&#39;][&#39;name&#39;],</span>
<span class="sd">                &#39;subject&#39;: template,</span>
<span class="sd">                &#39;query&#39;: integrant,</span>
<span class="sd">                &#39;subject_pos&#39;: &#39;{}:{}-{}&#39;.format(site[1][&#39;subject&#39;][&#39;name&#39;], left[0], right[1]),</span>
<span class="sd">                &#39;query_pos&#39;: &#39;{}:{}-{}&#39;.format(site[1][&#39;query&#39;][&#39;name&#39;], query_range[0], query_range[1]),</span>
<span class="sd">                &#39;left_homology_pos&#39;: &#39;{}:{}-{}&#39;.format(site[1][&#39;subject&#39;][&#39;name&#39;], left[0], left[1]),</span>
<span class="sd">                &#39;right_homology_pos&#39;: &#39;{}:{}-{}&#39;.format(site[1][&#39;subject&#39;][&#39;name&#39;], right[0], right[1]),</span>
<span class="sd">                &#39;left_homology_range&#39;: left,</span>
<span class="sd">                &#39;right_homology_range&#39;: right,</span>
<span class="sd">                &#39;flanking_bps&#39;: flanking_bps,</span>
<span class="sd">                &#39;flank_left_range&#39;: flank_left,</span>
<span class="sd">                &#39;flank_right_range&#39;: flank_right,</span>
<span class="sd">                &#39;query_range&#39;: query_range,</span>
<span class="sd">                &#39;sequence&#39;: {</span>
<span class="sd">                    &#39;flank_left&#39;: flank_left_seq,</span>
<span class="sd">                    &#39;left_homology&#39;: left_seq,</span>
<span class="sd">                    &#39;integrant&#39;: integrant_seq,</span>
<span class="sd">                    &#39;right_homology&#39;: right_seq,</span>
<span class="sd">                    &#39;flank_right&#39;: flank_right_seq</span>
<span class="sd">                },</span>
<span class="sd">                &#39;direction&#39;: direction,</span>
<span class="sd">                &#39;alignment&#39;: (site[1], site[2])</span>
<span class="sd">            }</span>


<span class="sd">        :param aligner: aligner</span>
<span class="sd">        :type aligner: pyblast.JSONBlast</span>
<span class="sd">        :param integrant: integrant sequence</span>
<span class="sd">        :type integrant: jdna.Sequence</span>
<span class="sd">        :param template: template sequence</span>
<span class="sd">        :type template: jdna.Sequence</span>
<span class="sd">        :param min_homology: minimum number of homologous bps to consider a valid recombination event (default: 100)</span>
<span class="sd">        :type min_homology: int</span>
<span class="sd">        :param flanking_bps: the number of outside flanking bps to return (default: 500)</span>
<span class="sd">        :type flanking_bps: int</span>
<span class="sd">        :param max_distance: maxmum distance (bps) between homologous regions to consider a valid recombination event (default: 10000)</span>
<span class="sd">        :type max_distance: int</span>
<span class="sd">        :return: list of integration events</span>
<span class="sd">        :rtype: list</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">alignments</span>
        <span class="n">sites</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">alignments_to_integration_sites</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
        <span class="n">valid_sites</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">invalid_sites</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">sites</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;alignment_length&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">min_homology</span><span class="p">:</span>
                <span class="n">invalid_sites</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                    <span class="s1">&#39;REASON&#39;</span><span class="p">:</span> <span class="s1">&#39;homology 1 less than min homology (</span><span class="si">{}</span><span class="s1"> &lt; </span><span class="si">{}</span><span class="s1">)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;alignment_length&#39;</span><span class="p">],</span>
                                                                                   <span class="n">min_homology</span><span class="p">),</span>
                    <span class="s1">&#39;SITE&#39;</span><span class="p">:</span> <span class="n">site</span>
                <span class="p">})</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">site</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;meta&#39;</span><span class="p">][</span><span class="s1">&#39;alignment_length&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">min_homology</span><span class="p">:</span>
                <span class="n">invalid_sites</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                    <span class="s1">&#39;REASON&#39;</span><span class="p">:</span> <span class="s1">&#39;homology 2 less than min homology (</span><span class="si">{}</span><span class="s1"> &lt; </span><span class="si">{}</span><span class="s1">)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;alignment_length&#39;</span><span class="p">],</span>
                                                                                   <span class="n">min_homology</span><span class="p">),</span>
                    <span class="s1">&#39;SITE&#39;</span><span class="p">:</span> <span class="n">site</span>
                <span class="p">})</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">&gt;</span> <span class="n">max_distance</span><span class="p">:</span>
                <span class="n">invalid_sites</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                    <span class="s1">&#39;REASON&#39;</span><span class="p">:</span> <span class="s1">&#39;max distance between homologies is too great (</span><span class="si">{}</span><span class="s1"> &gt; </span><span class="si">{}</span><span class="s1">)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
                                                                                              <span class="n">max_distance</span><span class="p">),</span>
                    <span class="s1">&#39;SITE&#39;</span><span class="p">:</span> <span class="n">site</span>
                <span class="p">})</span>
                <span class="k">continue</span>
            <span class="n">valid_sites</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">site</span><span class="p">)</span>

        <span class="n">integrations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">valid_sites</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">integrant</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">integrant</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">seq_dict</span><span class="p">[</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]][</span><span class="s1">&#39;bases&#39;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">template</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">template</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">seq_dict</span><span class="p">[</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;sequence_id&#39;</span><span class="p">]][</span><span class="s1">&#39;bases&#39;</span><span class="p">]</span>
            <span class="n">query_range</span> <span class="o">=</span> <span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">],</span> <span class="n">site</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">])</span>
            <span class="n">integrant_seq</span> <span class="o">=</span> <span class="n">integrant</span><span class="p">[</span><span class="n">query_range</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">query_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
            <span class="n">direction</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">site</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">left</span> <span class="o">=</span> <span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">])</span>
                <span class="n">right</span> <span class="o">=</span> <span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">site</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">integrant_seq</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
                <span class="n">direction</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                <span class="n">left</span> <span class="o">=</span> <span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">],</span> <span class="n">site</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">])</span>
                <span class="n">right</span> <span class="o">=</span> <span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">],</span> <span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">])</span>
            <span class="n">flank_left</span> <span class="o">=</span> <span class="p">(</span><span class="n">left</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">flanking_bps</span><span class="p">,</span> <span class="n">left</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">flank_right</span> <span class="o">=</span> <span class="p">(</span><span class="n">right</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">right</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">flanking_bps</span><span class="p">)</span>

            <span class="n">flank_left_seq</span> <span class="o">=</span> <span class="n">template</span><span class="p">[</span><span class="n">flank_left</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">flank_left</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
            <span class="n">left_seq</span> <span class="o">=</span> <span class="n">template</span><span class="p">[</span><span class="n">left</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">left</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
            <span class="n">right_seq</span> <span class="o">=</span> <span class="n">template</span><span class="p">[</span><span class="n">right</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">right</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
            <span class="n">flank_right_seq</span> <span class="o">=</span> <span class="n">template</span><span class="p">[</span><span class="n">flank_right</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">flank_right</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>

            <span class="n">integration_info</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;subject_name&#39;</span><span class="p">:</span> <span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span>
                <span class="s1">&#39;query_name&#39;</span><span class="p">:</span> <span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span>
                <span class="s1">&#39;subject&#39;</span><span class="p">:</span> <span class="n">template</span><span class="p">,</span>
                <span class="s1">&#39;query&#39;</span><span class="p">:</span> <span class="n">integrant</span><span class="p">,</span>
                <span class="s1">&#39;subject_pos&#39;</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">:</span><span class="si">{}</span><span class="s1">-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> <span class="n">left</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">right</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                <span class="s1">&#39;query_pos&#39;</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">:</span><span class="si">{}</span><span class="s1">-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;query&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> <span class="n">query_range</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">query_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                <span class="s1">&#39;left_homology_pos&#39;</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">:</span><span class="si">{}</span><span class="s1">-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> <span class="n">left</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">left</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                <span class="s1">&#39;right_homology_pos&#39;</span><span class="p">:</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">:</span><span class="si">{}</span><span class="s1">-</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;subject&#39;</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> <span class="n">right</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">right</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                <span class="s1">&#39;left_homology_range&#39;</span><span class="p">:</span> <span class="n">left</span><span class="p">,</span>
                <span class="s1">&#39;right_homology_range&#39;</span><span class="p">:</span> <span class="n">right</span><span class="p">,</span>
                <span class="s1">&#39;flanking_bps&#39;</span><span class="p">:</span> <span class="n">flanking_bps</span><span class="p">,</span>
                <span class="s1">&#39;flank_left_range&#39;</span><span class="p">:</span> <span class="n">flank_left</span><span class="p">,</span>
                <span class="s1">&#39;flank_right_range&#39;</span><span class="p">:</span> <span class="n">flank_right</span><span class="p">,</span>
                <span class="s1">&#39;query_range&#39;</span><span class="p">:</span> <span class="n">query_range</span><span class="p">,</span>
                <span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s1">&#39;flank_left&#39;</span><span class="p">:</span> <span class="n">flank_left_seq</span><span class="p">,</span>
                    <span class="s1">&#39;left_homology&#39;</span><span class="p">:</span> <span class="n">left_seq</span><span class="p">,</span>
                    <span class="s1">&#39;integrant&#39;</span><span class="p">:</span> <span class="n">integrant_seq</span><span class="p">,</span>
                    <span class="s1">&#39;right_homology&#39;</span><span class="p">:</span> <span class="n">right_seq</span><span class="p">,</span>
                    <span class="s1">&#39;flank_right&#39;</span><span class="p">:</span> <span class="n">flank_right_seq</span>
                <span class="p">},</span>
                <span class="s1">&#39;direction&#39;</span><span class="p">:</span> <span class="n">direction</span><span class="p">,</span>
                <span class="s1">&#39;alignment&#39;</span><span class="p">:</span> <span class="p">(</span><span class="n">site</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">site</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
            <span class="p">}</span>

            <span class="n">integrations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">integration_info</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">integrations</span></div>

<div class="viewcode-block" id="BlastUtils.integration_sequence"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.BlastUtils.integration_sequence">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">integration_sequence</span><span class="p">(</span><span class="n">integration_results</span><span class="p">:</span> <span class="nb">dict</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Parse an integration result into a :class:`jdna.Sequence` instance.&quot;&quot;&quot;</span>

        <span class="n">flank_left</span> <span class="o">=</span> <span class="n">jdna</span><span class="o">.</span><span class="n">Sequence</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="s1">&#39;flank_left&#39;</span><span class="p">]))</span>
        <span class="n">flank_left</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;subject_name&#39;</span><span class="p">])</span>

        <span class="n">left</span> <span class="o">=</span> <span class="n">jdna</span><span class="o">.</span><span class="n">Sequence</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="s1">&#39;left_homology&#39;</span><span class="p">]))</span>
        <span class="n">left</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;left_homology_pos&#39;</span><span class="p">])</span>

        <span class="n">flank_right</span> <span class="o">=</span> <span class="n">jdna</span><span class="o">.</span><span class="n">Sequence</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="s1">&#39;flank_right&#39;</span><span class="p">]))</span>
        <span class="n">flank_right</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;subject_name&#39;</span><span class="p">])</span>

        <span class="n">right</span> <span class="o">=</span> <span class="n">jdna</span><span class="o">.</span><span class="n">Sequence</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="s1">&#39;right_homology&#39;</span><span class="p">]))</span>
        <span class="n">right</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;right_homology_pos&#39;</span><span class="p">])</span>

        <span class="n">integration_seq</span> <span class="o">=</span> <span class="n">flank_left</span> <span class="o">+</span> <span class="n">left</span> <span class="o">+</span> <span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][</span><span class="s1">&#39;integrant&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">right</span> <span class="o">+</span> <span class="n">flank_right</span>

        <span class="k">return</span> <span class="n">integration_seq</span></div>

    <span class="c1"># TODO: allow a provided list of primers to select from.</span>
    <span class="c1"># TODO: allow a design of left or right primers</span>
<div class="viewcode-block" id="BlastUtils.integration_qc_primers"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.BlastUtils.integration_qc_primers">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">integration_qc_primers</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">integration_results</span><span class="p">,</span> <span class="n">size_range</span><span class="p">,</span> <span class="n">session</span><span class="p">:</span> <span class="n">AqSession</span><span class="p">,</span> <span class="n">pick_left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                               <span class="n">pick_right</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Parse an integration result into a :class:`jdna.Sequence` instance and select PCR primers</span>
<span class="sd">        from the provided Aquarium session.</span>

<span class="sd">        :param integration_results: integration result</span>
<span class="sd">        :type integration_results: dict</span>
<span class="sd">        :param size_range: size range of pcr products</span>
<span class="sd">        :type size_range: tuple</span>
<span class="sd">        :param session: session</span>
<span class="sd">        :type session: AqSession</span>
<span class="sd">        :param pick_left: whether to pick left primers</span>
<span class="sd">        :type pick_left: bool</span>
<span class="sd">        :param pick_right: whether to pick right primers</span>
<span class="sd">        :type pick_right: bool</span>
<span class="sd">        :return: (seq, left_qc_primers, right_qc_primers)</span>
<span class="sd">        :rtype: tuple</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">integration_seq</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">integration_sequence</span><span class="p">(</span><span class="n">integration_results</span><span class="p">)</span>
        <span class="n">flanking_bps</span> <span class="o">=</span> <span class="n">integration_results</span><span class="p">[</span><span class="s1">&#39;flanking_bps&#39;</span><span class="p">]</span>

        <span class="n">left_qc_primers</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">right_qc_primers</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">pick_left</span><span class="p">:</span>
            <span class="n">left_target</span> <span class="o">=</span> <span class="p">(</span><span class="n">flanking_bps</span> <span class="o">-</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
            <span class="n">left_qc_primers</span> <span class="o">=</span> <span class="n">pick_pcr_primers</span><span class="p">(</span><span class="n">session</span><span class="p">,</span> <span class="n">integration_seq</span><span class="p">[:</span><span class="n">flanking_bps</span> <span class="o">*</span> <span class="mi">2</span><span class="p">],</span> <span class="n">size_range</span><span class="p">,</span>
                                               <span class="n">target</span><span class="o">=</span><span class="n">left_target</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">pick_right</span><span class="p">:</span>
            <span class="n">right_target</span> <span class="o">=</span> <span class="p">(</span><span class="n">flanking_bps</span> <span class="o">-</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
            <span class="n">right_qc_primers</span> <span class="o">=</span> <span class="n">pick_pcr_primers</span><span class="p">(</span><span class="n">session</span><span class="p">,</span> <span class="n">integration_seq</span><span class="p">[</span><span class="o">-</span><span class="n">flanking_bps</span> <span class="o">*</span> <span class="mi">2</span><span class="p">:],</span> <span class="n">size_range</span><span class="p">,</span>
                                                <span class="n">target</span><span class="o">=</span><span class="n">right_target</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">integration_seq</span><span class="p">,</span> <span class="n">left_qc_primers</span><span class="p">,</span> <span class="n">right_qc_primers</span></div>


<div class="viewcode-block" id="BlastUtils.homologous_recombination"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.BlastUtils.homologous_recombination">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">homologous_recombination</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">integrant</span><span class="p">:</span> <span class="n">jdna</span><span class="o">.</span><span class="n">Sequence</span><span class="p">,</span> <span class="n">template</span><span class="p">:</span> <span class="n">jdna</span><span class="o">.</span><span class="n">Sequence</span><span class="p">,</span> <span class="n">min_homology</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flanking_bps</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
                                 <span class="n">max_distance</span><span class="o">=</span><span class="mi">5000</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parse a **run** :class:`pyblast.JSONBlast` instance into valid integration sites. Returns the following info</span>
<span class="sd">        for each integration:</span>

<span class="sd">        .. code-block::</span>

<span class="sd">            integration_info = {</span>
<span class="sd">                &#39;subject_name&#39;: site[1][&#39;subject&#39;][&#39;name&#39;],</span>
<span class="sd">                &#39;query_name&#39;: site[1][&#39;query&#39;][&#39;name&#39;],</span>
<span class="sd">                &#39;subject&#39;: template,</span>
<span class="sd">                &#39;query&#39;: integrant,</span>
<span class="sd">                &#39;subject_pos&#39;: &#39;{}:{}-{}&#39;.format(site[1][&#39;subject&#39;][&#39;name&#39;], left[0], right[1]),</span>
<span class="sd">                &#39;query_pos&#39;: &#39;{}:{}-{}&#39;.format(site[1][&#39;query&#39;][&#39;name&#39;], query_range[0], query_range[1]),</span>
<span class="sd">                &#39;left_homology_pos&#39;: &#39;{}:{}-{}&#39;.format(site[1][&#39;subject&#39;][&#39;name&#39;], left[0], left[1]),</span>
<span class="sd">                &#39;right_homology_pos&#39;: &#39;{}:{}-{}&#39;.format(site[1][&#39;subject&#39;][&#39;name&#39;], right[0], right[1]),</span>
<span class="sd">                &#39;left_homology_range&#39;: left,</span>
<span class="sd">                &#39;right_homology_range&#39;: right,</span>
<span class="sd">                &#39;flanking_bps&#39;: flanking_bps,</span>
<span class="sd">                &#39;flank_left_range&#39;: flank_left,</span>
<span class="sd">                &#39;flank_right_range&#39;: flank_right,</span>
<span class="sd">                &#39;query_range&#39;: query_range,</span>
<span class="sd">                &#39;sequence&#39;: {</span>
<span class="sd">                    &#39;flank_left&#39;: flank_left_seq,</span>
<span class="sd">                    &#39;left_homology&#39;: left_seq,</span>
<span class="sd">                    &#39;integrant&#39;: integrant_seq,</span>
<span class="sd">                    &#39;right_homology&#39;: right_seq,</span>
<span class="sd">                    &#39;flank_right&#39;: flank_right_seq</span>
<span class="sd">                },</span>
<span class="sd">                &#39;direction&#39;: direction,</span>
<span class="sd">                &#39;alignment&#39;: (site[1], site[2])</span>
<span class="sd">            }</span>


<span class="sd">        :param integrant: integrant sequence</span>
<span class="sd">        :type integrant: jdna.Sequence</span>
<span class="sd">        :param template: template sequence</span>
<span class="sd">        :type template: jdna.Sequence</span>
<span class="sd">        :param min_homology: minimum number of homologous bps to consider a valid recombination event (default: 100)</span>
<span class="sd">        :type min_homology: int</span>
<span class="sd">        :param flanking_bps: the number of outside flanking bps to return (default: 500)</span>
<span class="sd">        :type flanking_bps: int</span>
<span class="sd">        :param max_distance: maxmum distance (bps) between homologous regions to consider a valid recombination event (default: 10000)</span>
<span class="sd">        :type max_distance: int</span>
<span class="sd">        :return: list of integration events</span>
<span class="sd">        :rtype: list</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">alignments</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">align</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">integrant</span><span class="p">)],</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">template</span><span class="p">)])</span>
        <span class="n">results</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">aligner_to_integrations</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">integrant</span><span class="o">=</span><span class="n">integrant</span><span class="p">,</span> <span class="n">template</span><span class="o">=</span><span class="n">template</span><span class="p">,</span>
                                           <span class="n">min_homology</span><span class="o">=</span><span class="n">min_homology</span><span class="p">,</span>
                                           <span class="n">flanking_bps</span><span class="o">=</span><span class="n">flanking_bps</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="n">max_distance</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span></div></div>

<div class="viewcode-block" id="Genome"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.Genome">[docs]</a><span class="k">class</span> <span class="nc">Genome</span><span class="p">(</span><span class="n">BlastUtils</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Genome utilities&quot;&quot;&quot;</span>

    <span class="n">genome_name</span> <span class="o">=</span> <span class="s1">&#39;Unknown&#39;</span>
    <span class="n">service_url</span> <span class="o">=</span> <span class="s2">&quot;interminewebservice&quot;</span>
    <span class="n">service</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">default_genome_path</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="Genome.chromosomes"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.Genome.chromosomes">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">chromosomes</span><span class="p">(</span><span class="bp">cls</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a list of chromosomes&quot;&quot;&quot;</span>
        <span class="n">query</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">service</span><span class="o">.</span><span class="n">new_query</span><span class="p">(</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">)</span>
        <span class="n">query</span><span class="o">.</span><span class="n">add_view</span><span class="p">(</span>
            <span class="s2">&quot;primaryIdentifier&quot;</span><span class="p">,</span>
            <span class="s2">&quot;featureType&quot;</span><span class="p">,</span> <span class="s2">&quot;length&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">query</span></div>

<div class="viewcode-block" id="Genome._chromosome_sequence"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.Genome._chromosome_sequence">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">_chromosome_sequence</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">identifier</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the sequence of a specific chromosome</span>

<span class="sd">        :param identifier: chromosome identifier</span>
<span class="sd">        :type identifier: basestring</span>
<span class="sd">        :return:</span>
<span class="sd">        :rtype:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">query</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">service</span><span class="o">.</span><span class="n">new_query</span><span class="p">(</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">)</span>
        <span class="n">query</span><span class="o">.</span><span class="n">add_view</span><span class="p">(</span>
            <span class="s1">&#39;primaryIdentifier&#39;</span><span class="p">,</span> <span class="s1">&#39;length&#39;</span><span class="p">,</span> <span class="s1">&#39;featureType&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence.residues&#39;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">identifier</span><span class="p">:</span>
            <span class="n">query</span><span class="o">.</span><span class="n">add_constraint</span><span class="p">(</span><span class="s2">&quot;primaryIdentifier&quot;</span><span class="p">,</span> <span class="s2">&quot;=&quot;</span><span class="p">,</span> <span class="n">identifier</span><span class="p">,</span> <span class="n">code</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">query</span></div>

<div class="viewcode-block" id="Genome.genome"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.Genome.genome">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">genome</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">filepath</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return genome.</span>

<span class="sd">        :param filepath: filepath to save the genome</span>
<span class="sd">        :type filepath: basestring</span>
<span class="sd">        :param overwrite: if True, overwrite existing file</span>
<span class="sd">        :type overwrite: bool</span>
<span class="sd">        :return: list of SeqRecords</span>
<span class="sd">        :rtype: list</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">filepath</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">filepath</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">default_genome_path</span>
        <span class="n">filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">overwrite</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
            <span class="n">rows</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_chromosome_sequence</span><span class="p">()</span><span class="o">.</span><span class="n">rows</span><span class="p">()</span>
            <span class="n">seqrecords</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rows</span><span class="p">:</span>
                <span class="n">seq</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;sequence.residues&#39;</span><span class="p">]),</span> <span class="nb">id</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;primaryIdentifier&#39;</span><span class="p">])</span>
                <span class="n">seqrecords</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seqrecords</span><span class="p">,</span> <span class="n">filepath</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">seqrecords</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">))</span></div>

<div class="viewcode-block" id="Genome.genome_json"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.Genome.genome_json">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">genome_json</span><span class="p">(</span><span class="bp">cls</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return json object representing the genome&quot;&quot;&quot;</span>
        <span class="n">seqs</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">genome</span><span class="p">()</span>
        <span class="k">return</span> <span class="p">[{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="s1">&#39;bases&#39;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span>
            <span class="s1">&#39;circular&#39;</span><span class="p">:</span> <span class="kc">False</span>
        <span class="p">}</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">]</span></div>

<div class="viewcode-block" id="Genome.chromosome"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.Genome.chromosome">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">chromosome</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">iden</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a chromosome&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">cls</span><span class="o">.</span><span class="n">genome</span><span class="p">()[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">seq</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">iden</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">seq</span></div>

<div class="viewcode-block" id="Genome.genome_aligner"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.Genome.genome_aligner">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">genome_aligner</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">sequences</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return a blast aligner to the genome</span>

<span class="sd">        :param sequences: list of sequences to align to genome</span>
<span class="sd">        :type sequences: list</span>
<span class="sd">        :return: aligner</span>
<span class="sd">        :rtype: pyblast.JSONBlast</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span>

        <span class="n">query_json</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="n">query_json</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                    <span class="s1">&#39;bases&#39;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="p">),</span>
                    <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;query&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;circular&#39;</span><span class="p">:</span> <span class="kc">False</span>
                <span class="p">})</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">jdna</span><span class="o">.</span><span class="n">Sequence</span><span class="p">):</span>
                <span class="n">query_json</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                    <span class="s1">&#39;bases&#39;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="p">),</span>
                    <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="n">seq</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                    <span class="s1">&#39;circular&#39;</span><span class="p">:</span> <span class="kc">False</span>
                <span class="p">})</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;Type </span><span class="si">{}</span><span class="s2"> not recognized&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="p">)))</span>

        <span class="n">aligner</span> <span class="o">=</span> <span class="n">pyblast</span><span class="o">.</span><span class="n">JSONBlast</span><span class="p">(</span><span class="n">query_json</span><span class="o">=</span><span class="n">query_json</span><span class="p">,</span> <span class="n">subject_json</span><span class="o">=</span><span class="bp">cls</span><span class="o">.</span><span class="n">genome_json</span><span class="p">(),</span> <span class="n">span_origin</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">aligner</span></div>

<div class="viewcode-block" id="Genome.align_to_genome"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.Genome.align_to_genome">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">align_to_genome</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">sequences</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Align sequences to genome using blast.</span>

<span class="sd">        :param sequences: list of sequences</span>
<span class="sd">        :type sequences: list</span>
<span class="sd">        :return: aligner</span>
<span class="sd">        :rtype: pyblast.JSONBlast</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">aligner</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">genome_aligner</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
        <span class="n">aligner</span><span class="o">.</span><span class="n">quick_blastn</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">aligner</span></div>

<div class="viewcode-block" id="Genome.genomic_integration"><a class="viewcode-back" href="../../../developer/_autosummary/_autosummary/tridentplus.organisms.genome.html#tridentplus.organisms.genome.Genome.genomic_integration">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">genomic_integration</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">integrant</span><span class="p">:</span> <span class="n">jdna</span><span class="o">.</span><span class="n">Sequence</span><span class="p">,</span> <span class="n">min_homology</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">flanking_bps</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="mi">5000</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Simulate a genomic integration event.</span>

<span class="sd">        :param integrant: integrant sequence</span>
<span class="sd">        :type integrant: jdna.Sequence</span>
<span class="sd">        :param min_homology: minimum homology to consider</span>
<span class="sd">        :type min_homology: int</span>
<span class="sd">        :param flanking_bps:</span>
<span class="sd">        :type flanking_bps:</span>
<span class="sd">        :param max_distance:</span>
<span class="sd">        :type max_distance:</span>
<span class="sd">        :return:</span>
<span class="sd">        :rtype:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">alignments</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">align_to_genome</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">integrant</span><span class="p">)])</span>
        <span class="n">results</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">aligner_to_integrations</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">integrant</span><span class="o">=</span><span class="n">integrant</span><span class="p">,</span> <span class="n">min_homology</span><span class="o">=</span><span class="n">min_homology</span><span class="p">,</span>
                                           <span class="n">flanking_bps</span><span class="o">=</span><span class="n">flanking_bps</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="n">max_distance</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span></div></div>
</pre></div>

          </div>
            
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">TridentPlus 0.0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="../../../_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2017, University of Washington. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>